src/ray/object_manager/pull_manager.cc

```diff
@@ -74,58 +74,73 @@ void PullManager::OnLocationChange(const ObjectID &object_id,
   // we may end up sending a duplicate request to the same client as
   // before.
   it->second.client_locations = std::vector<NodeID>(client_ids.begin(), client_ids.end());
+
   it->second.spilled_url = spilled_url;
   RAY_LOG(DEBUG) << "OnLocationChange " << spilled_url << " num clients "
                  << client_ids.size();
 
   TryToMakeObjectLocal(object_id);
 }
 
-void PullManager::TryToMakeObjectLocal(const ObjectID &object_id) {
+void PullManager::TryToMakeObjectLocal(const ObjectID &object_id, bool resend) {
   if (object_is_local_(object_id)) {
+    RAY_LOG(ERROR) << "object is already local";
     return;
   }
   auto it = object_pull_requests_.find(object_id);
   if (it == object_pull_requests_.end()) {
+    RAY_LOG(ERROR) << "Couldn't find pr";
     return;
   }
 
   auto &request = it->second;
-  if (!request.spilled_url.empty()) {
-    // Try to restore the spilled object.
-    restore_spilled_object_(object_id, request.spilled_url,
-                            [this, object_id](const ray::Status &status) {
-                              // Fall back to fetching from another object manager.
-                              if (!status.ok()) {
-                                PullFromRandomLocation(object_id);
-                              }
-                            });
-  } else {
-    // New object locations were found, so begin trying to pull from a
-    // client. This will be called every time a new client location
-    // appears.
-    PullFromRandomLocation(object_id);
+  bool sent_pull = false;
+
+  const auto &spilled_url = request.spilled_url;
+  if (request.inflight_pulls.size() == 0 || resend) {
+    if (!spilled_url.empty()) {
+      RAY_LOG(ERROR) << "Pulling from spilled location.";
+      RAY_LOG(DEBUG) << "OnLocationChange " << spilled_url << " num clients "
+                     << request.client_locations.size();
+      // Try to restore the spilled object.
+      restore_spilled_object_(object_id, spilled_url,
+                              [this, object_id](const ray::Status &status) {
+                                // Fall back to fetching from another object manager.
+                                if (!status.ok()) {
+                                  PullFromRandomLocation(object_id);
+                                }
+                              });
+    } else {
+      // New object locations were found, so begin trying to pull from a
+      // client. This will be called every time a new client location
+      // appears.
+      sent_pull = PullFromRandomLocation(object_id);
+    }
   }
 
-  const auto time = get_time_();
-  auto retry_timeout_len = (pull_timeout_ms_ / 1000.) * (1UL << request.num_retries);
-  request.next_pull_time = time + retry_timeout_len;
+  if (sent_pull) {
+    const auto time = get_time_();
```

Comment:
I think it's a bit more readable to move this code into the above if-else or PullFromRandomLocation instead of adding the send_pull bool. Or we could just always reset the timer, even if we didn't send a pull request (it should be cheap to check whether we need to resend).